### PR TITLE
Read black levels from Nikon makernotes

### PR DESCRIPTION
**Pull request by kmilos**
Using https://exiftool.org/TagNames/Nikon.html

### CODE REVIEW

File: src/librawspeed/decoders/NefDecoder.cpp [hunk 1]
```diff
@@ -607,6 +607,19 @@ void NefDecoder::decodeMetaDataInternal(const CameraMetaData* meta) {
     setMetaData(meta, id, "", iso);
   }
 
+  // Read any additional black levels, but only if not already overriden
+  if (!hints.has("override_cfa_black") &&
```
Comment (LebedevRI) [inline src/librawspeed/decoders/NefDecoder.cpp]: This should not be here.

File: src/librawspeed/decoders/NefDecoder.cpp [hunk 2]
```diff
@@ -607,6 +607,19 @@ void NefDecoder::decodeMetaDataInternal(const CameraMetaData* meta) {
...
+      mRootIFD->hasEntryRecursive(static_cast<TiffTag>(0x003d))) {
```
Comment (LebedevRI) [inline src/librawspeed/decoders/NefDecoder.cpp]: Please use named Tiff Tag enum, add one if it's not there yet.

**Comment by kmilos** [comment]
Ah, this won't work, as I just read for the D5600 samples (ISO 200)

```
BlackLevel                      : 600 600 600 600
```

in both 12-bit and 14-bit modes...

Converting to draft until someone comes up with an additional fix...

**Comment by kmilos** [comment]
One possibility is just going by `mode`.

**Comment by LebedevRI** [comment]
```c++
uint32_t bitPerPixel = raw->getEntry(TiffTag::BITSPERSAMPLE)->getU32();
if(bitPerPixel != 12 && bitPerPixel != 14)
      ThrowRDE("Bad bit per pixel: ", bitPerPixel);
unsigned shamt = 14-bitPerPixel;
for(auto&b : mRaw->blackLevelSeparate)
  b >>= shamt;
mRaw->whiteLevel >>= shamt;
```

**Comment by LebedevRI** [comment]
Looking at the results, not really sure how much this helps anything, but it's certainly a step in the right direction.
Thank you!

**Comment by kmilos** [comment]
Thanks. Indeed, in limited testing I haven't yet come across a sample where the values are different from the ones in cameras.xml...

**Pull request merged**